File: csharp/DiagramPrinter/PrinterDriverFactory.cs
using System.Runtime.CompilerServices;

[assembly: InternalsVisibleTo("DiagramPrinterTests")]

namespace DiagramPrinter;

public class PrinterDriverFactory
{
    private static PrinterDriverFactory _instance;

    public static PrinterDriverFactory Instance
    {
        get
        {
            if (_instance == null)
            {
                _instance = new PrinterDriverFactory();
            }

            return _instance;
        }
        // for unit tests
        internal set => _instance = value;
    }

    public virtual DiagramPrintDriver CreateDriverForPrint()
    {
        return new DiagramPrintDriver();
    }
}

File: java/src/main/java/sammancoaching/PrinterDriverFactory.java
package sammancoaching;

public class PrinterDriverFactory {
    private static PrinterDriverFactory instance;

    public static PrinterDriverFactory getInstance() {
        if (instance == null) {
            instance = new PrinterDriverFactory();
        }
        return instance;
    }

    // For unit tests
    static void setInstance(PrinterDriverFactory newInstance) {
        instance = newInstance;
    }

    public DiagramPrintDriver createDriverForPrint() {
        return new DiagramPrintDriver();
    }
}
